relay/shfu.go

```diff
@@ -0,0 +1,149 @@
+package relay
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/avast/retry-go"
+	"github.com/cosmos/cosmos-sdk/codec/types"
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+	ibcexported "github.com/cosmos/ibc-go/v8/modules/core/exported"
+	"github.com/datachainlab/lcp-go/relay/shfu_logger"
+	"github.com/datachainlab/lcp-go/relay/shfu_storage"
+	"github.com/hyperledger-labs/yui-relayer/core"
+	"github.com/hyperledger-labs/yui-relayer/coreutil"
+	"github.com/hyperledger-labs/yui-relayer/log"
+)
+
+// Re-export types from storage package for backward compatibility
+type SHFURecord = shfu_storage.SHFURecord
+type SHFUStorage = shfu_storage.SHFUStorage
+
+// getSHFULogger gets logger for SHFU operations, first from context, then from prover
+func getSHFULogger(ctx context.Context, target *core.ProvableChain) *log.RelayLogger {
+	// First try to get logger from context
+	if logger := shfu_logger.GetSHFULoggerOrNil(ctx); logger != nil {
+		return logger
+	}
+
+	// If no logger in context, try to get from prover
+	if lcpProver, err := coreutil.UnwrapProver[*Prover](target.Prover); err == nil {
+		return lcpProver.getLogger()
+	}
+
+	// Fallback to default SHFU logger
+	return shfu_logger.GetSHFULogger(ctx)
+}
+
+// ExecuteSetupHeadersForUpdate executes SetupHeadersForUpdate0 and returns UpdateClientResult array
+// This function can be used by various commands and services
+// fromHeight: the starting height for SHFU operations (nil for unspecified)
+// counterparty: the counterparty chain object
+func SHFUExecuteAndStore(ctx context.Context, target *core.ProvableChain, counterparty *core.ProvableChain, storage SHFUStorage) (*SHFURecord, error) {
+	// Get SHFU logger once and reuse throughout the function
+	logger := getSHFULogger(ctx, target)
+
+	// Check if counterparty chain is provided
+	if counterparty == nil {
+		return nil, fmt.Errorf("counterparty chain is required for SHFU operations")
+	}
+
+	// Unwrap LCP prover from the chain to get configuration
+	lcpProver, err := coreutil.UnwrapProver[*Prover](target.Prover)
+	if err != nil {
+		return nil, fmt.Errorf("chain %q is not an LCP prover: %w", target.ChainID(), err)
+	}
+
+	// Try to get the finalized header
+	latestFinalizedHeader, err := target.GetLatestFinalizedHeader(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get latest finalized header: %w", err)
+	}
+
+	toHeight := latestFinalizedHeader.GetHeight()
+
+	// Check if we already have a record with toHeight >= current toHeight
+	latestRecord, err := storage.GetLatestSHFUForChain(ctx, target.ChainID(), counterparty.ChainID())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get latest SHFU record: %w", err)
+	}
+	if latestRecord != nil && !latestRecord.ToHeight.LT(toHeight) {
```

Comment:
This condition is a little confusing:

```suggestion
	if latestRecord != nil && latestRecord.ToHeight.GTE(toHeight) {
```

Reply:
not to limit by toHeight

Reply:
Sorry, what do you mean by "not to limit by toHeight"?
I think `latestRecord.ToHeight.GTE(toHeight)` is identical to `!latestRecord.ToHeight.LT(toHeight)` and more readable.
Anyway, this line appear to have been removed in https://github.com/datachainlab/lcp-go/pull/58/changes/5f526297b172df22f11dcfccdb8f75cd8ad6d15c.